---
title: "Finemapping of ld regions using single trait Bayesian Linear Regression models"
author: "Peter SÃ¸rensen"
date: "`r Sys.Date()`"
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
This script is focusing on fine-mapping using the Bayesian Linear Regression (BLR) model. It processes and analyzes GWAS (Genome-Wide Association Studies) summary statistics and employs fine-mapping on specific ld regions.

```{r all-code, include=TRUE, eval=FALSE}
# Load libraries
library(qgg)
library(gact)
library(data.table)

# Load GAlist
GAlist <- readRDS(file="C:/Users/gact/hsa.0.0.1/GAlist_hsa.0.0.1.rds")

# Save Glist for use later
Glist <- readRDS(file=file.path(GAlist$dirs["marker"],"Glist_1000G.rds"))


# Load Glist with information on 1000G
Glist <- readRDS(file=file.path(GAlist$dirs["marker"],"Glist_1000G.rds"))

# Select study
studyID <- "GWAS1"

# Extract GWAS summary statistics for GWAS1
stat <- getMarkerStatDB(GAlist=GAlist, studyID=studyID)

# Check and align summary statistics based on marker information in Glist
stat <- checkStat(Glist=Glist, stat=stat,
                  excludeMAF=0.05,
                  excludeMAFDIFF=0.05,
                  excludeINFO=0.8,
                  excludeCGAT=TRUE,
                  excludeINDEL=TRUE,
                  excludeDUPS=TRUE,
                  excludeMHC=FALSE,
                  excludeMISS=0.05,
                  excludeHWE=1e-12)

# Fit BLR model across all chromosomes
for ( chr in 1:22) {

  # Create LD sets  
  sets <- qgg:::createLDsets(ldscores=Glist$ldscores[[chr]], 
                             maxsize=2000, msize=50, verbose=TRUE)
  
  # Map LD sets to rsids in stat
  sets <- qgg:::mapSets(sets, rsids=stat$rsids, index=FALSE)

  
  # Fine-mapping using BLR model on selected regions
  fit <- gmap(Glist=Glist, stat=stat, sets=sets,
              method="bayesC", pi=0.001, h2=0.1,
              vb=0.01, ssb_prior=0.01, nub=5,
              checkConvergence=TRUE, ntrial=3, 
              pruneLD=TRUE, r2=0.95, checkLD=FALSE,
              nit=1000, nburn=500, verbose=TRUE,
              updateB=TRUE, updatePi=FALSE, updateG=TRUE, updateE=TRUE)
  
  # Posterior estimates of hyper-parameters (ve,vg,vb,pi,pip) for every fine-mapped region
  head(fit$post)
  
  # Convergence statistics for every fine-mapped region
  head(fit$conv)
  
  # Posterior estimates of marker effects for every fine-mapped region
  head(fit$stat)
  
  # Save fit object for all sets on chromosome in directory of choice (here gbayes folder in database)
  saveRDS(fit, file=file.path(GAlist$dirs["gbayes"],paste0("fit_blr_chr",chr,"_",studyID,".rds")))
  
}
