---
title: "Pathway prioritization using a Bayesian MAGMA model and gene-level statistics from VEGAS"
author: "Peter SÃ¸rensen"
date: "`r Sys.Date()`"
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
This script is designed for advanced gene set analysis, utilizing a variety of tools and methodologies. Its primary focus lies in the integration and assessment of gene-marker sets, along with a detailed exploration of gene-disease associations derived from text mining, knowledge bases, and experimental data. The analysis utilizes gene-marker relationships, gene-level statistics, and gene-disease associations to conduct comprehensive gene set evaluations. This allow an in-depth analysis helping interpreting and validating genomic associations.

<br>
<div style="text-align: center;">

![](../Images/Bayesian_MAGMA.png){width=75%}

</div>

<br>



```{r all-code, include=TRUE, eval=FALSE}
# Load libraries
library(qgg)
library(gact)


# Load GAlist with information on gact database
GAlist <- readRDS(file="C:/Users/gact/hsa.0.0.1/GAlist_hsa.0.0.1.rds")

# Select GWAS study IDs
studyIDs <- c("GWAS1","GWAS2")

# Get VEGAS gene-level statistics (i.e. z-statistics)
stat <- getVEGAS(GAlist=GAlist, studyID=studyIDs)
head(stat)

# Get gene sets for KEGG pathways
sets <- getFeatureSets(GAlist = GAlist, feature="KEGG", minsets=50)


# MAGMA based on linear model  using a joint test for features (default MAGMA method)
fit <- magma(stat=stat[, "GWAS1",drop=FALSE], sets=sets, type="joint")
head(fit)

# MAGMA based on linear model using a marginal test for each feature
fit <- magma(stat=stat[, "GWAS1",drop=FALSE], sets=sets, type="marginal")
head(fit)

# Bayesian MAGMA based on single trait BLR model
fit <- magma(stat=stat[, "GWAS1",drop=FALSE], sets=sets, method="bayesC",
             pi=0.01, nit=5000, nburn=1000)
head(fit)


# Bayesian MAGMA based on a multiple trait BLR model
fit <- magma(stat=stat, sets=sets, method="bayesC",
             pi=0.01, nit=1000, nburn=200)
str(fit)
head(fit$PIP)

# MAGMA based on blr model (multiple trait for EUR, EAS, SAS T2D)
studyIDs <- c("GWAS6","GWAS7","GWAS8")
stat <- getVEGAS(GAlist=GAlist, studyID=studyIDs)
head(stat)
fit <- magma(stat=stat, sets=sets, method="bayesC", pi=0.01, nit=1000, nburn=200)
str(fit)
head(fit$b)
head(fit$PIP)


# MAGMA based on blr model (multiple trait for a range of T2D related traits)
studyIDs <- c("GWAS2", "GWAS6", "GWAS9", "GWAS12","GWAS13", "GWAS14", "GWAS16",  "GWAS17", "GWAS24", "GWAS25")
stat <- getVEGAS(GAlist=GAlist, studyID=studyIDs)
fit <- magma(stat=stat, sets=sets, method="bayesC",  pi=0.01, nit=1000, nburn=200)
str(fit)
head(fit$b)
head(fit$PIP)

